### PR TITLE
Rename controller _filters to _callbacks in docs

### PR DESCRIPTION
**Pull request by bf4**
Just docs update to remove references to controller `filters`.  I replaced `filter` with `callback`, where it seemed appropriate.  Perhaps `action callback` or `action lifecycle hooks` would be more appropriate?

### CODE REVIEW

**Comment by bf4** [comment]
Presumably these docs changes did not cause actionpack failures :) https://travis-ci.org/rails/rails/jobs/132127369

File: actionpack/lib/action_controller/metal/instrumentation.rb [hunk 1]
```diff
@@ -71,8 +71,8 @@ def redirect_to(*args)
   private
 
     # A hook invoked every time a before callback is halted.
-    def halted_callback_hook(filter)
-      ActiveSupport::Notifications.instrument("halted_callback.action_controller", :filter => filter)
+    def halted_callback_hook(callback)
+      ActiveSupport::Notifications.instrument("halted_callback.action_controller", callback: callback)
```
Comment (bf4) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: https://travis-ci.org/rails/rails/jobs/132127369.  looks like i missed wherever `payload[:filter]` is being called.  This change is obviously more invasive than the docs, so I leave that up to core if it is worth it.

Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: Probably not worth to change this. But all documentation changes are fine.

Comment (bf4) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: updated with just a comment

**Comment by bf4** [comment]
updated per feedback

**Comment by bf4** [comment]
@rafaelfranca any chance this will be in 5.0.0?

**Comment by bf4** [comment]
Related https://github.com/rails/rails/pull/18410

**Comment by bf4** [comment]
@rafaelfranca @maclover7 Any desire for me to update this? Or close it?

File: actionpack/lib/action_controller/metal/instrumentation.rb [hunk 2]
```diff
@@ -71,6 +71,7 @@ def redirect_to(*args)
   private
 
     # A hook invoked every time a before callback is halted.
+    # TODO: rename payload[:filter] to payload[:callback]
```
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: This will be really hard to rename since we can't deprecate since it is just a normal hash. Not sure if it is worth to rename. Maybe post on both keys?

**Review by rafaelfranca — COMMENTED** [review]
LGTM. I added a comment in the TODO note.

File: actionpack/lib/action_controller/metal/instrumentation.rb (hunk 2 above)
Comment (maclover7) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: 👍 on just leaving it. Deprecation cycle would be pretty tough 😞

**Comment by maclover7** [comment]
@bf4 Can you rebase on master, it looks like there are some merge conflicts

File: actionpack/lib/action_controller/metal/instrumentation.rb (hunk 2 above)
Comment (bf4) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: Will go rebase now

Comment (bf4) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: Do you think this deserves a comment? like

> # <tt>payload[:filter]</tt> refers to action callbacks previously being call filters.

Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: Yeah. A comment would be fine

**Comment by bf4** [comment]
@maclover7 @rafaelfranca sorry for the week-long delay in rebasing from when I said I would :(  Ready for review.  I'm going to point out some sections I have questions about.

File: actionpack/lib/action_controller/log_subscriber.rb [hunk 3]
```diff
@@ -32,6 +32,7 @@ def process_action(event)
       end
     end
 
+    # <tt>payload[:filter]</tt> refers to action callbacks, previously being call 'filters'.
     def halted_callback(event)
       info { "Filter chain halted as #{event.payload[:filter].inspect} rendered or redirected" }
```
Comment (bf4) [inline actionpack/lib/action_controller/log_subscriber.rb]: Is this helpful? Correct comment syntax?

File: actionpack/lib/action_controller/metal/instrumentation.rb [hunk 4]
```diff
@@ -71,6 +71,7 @@ def redirect_to(*args)
   private
 
     # A hook invoked every time a before callback is halted.
+    # <tt>payload[:filter]</tt> refers to action callbacks, previously being call 'filters'.
     def halted_callback_hook(filter)
       ActiveSupport::Notifications.instrument("halted_callback.action_controller", filter: filter)
```
Comment (bf4) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: Is this helpful? Correct comment syntax?

File: actionpack/test/controller/filters_test.rb [hunk 5]
```diff
@@ -15,6 +15,7 @@ def before_actions
   end
 end
 
+# <tt>filters</tt> refers to action callbacks, previously being call 'filters'.
```
Comment (bf4) [inline actionpack/test/controller/filters_test.rb]: Is this helpful? Correct comment syntax?  My intent is to call attention to the entire test file and a number of  variables still refer to action callbacks as `filters`.  I assume that any renaming of the file or variables would be done in a separate PR.

File: guides/source/5_1_release_notes.md [hunk 6]
```diff
@@ -292,7 +292,7 @@ Please refer to the [Changelog][action-mailer] for detailed changes.
 *   Allowed passing lambdas as values to the `default` method.
     ([Commit](https://github.com/rails/rails/commit/1cec84ad2ddd843484ed40b1eb7492063ce71baf))
 
-*   Added support for parameterized invocation of mailers to share before filters and defaults
+*   Added support for parameterized invocation of mailers to share before callbacks and defaults
```
Comment (bf4) [inline guides/source/5_1_release_notes.md]: Changed release note to be consistent with overall language as well as the before_action in the PR itself, even though David wrote it :)

File: guides/source/action_controller_overview.md [hunk 7]
```diff
@@ -10,7 +10,7 @@ After reading this guide, you will know:
 * How to follow the flow of a request through a controller.
 * How to restrict parameters passed to your controller.
 * How and why to store data in the session or cookies.
-* How to work with filters to execute code during request processing.
+* How to work with callbacks to execute code during request processing.
```
Comment (bf4) [inline guides/source/action_controller_overview.md]: perhaps instead of replacing `filters` with `callbacks` it should be `action callbacks`? Here and elsewhere

**Review by bf4 — COMMENTED** [review]
Some questions on areas that could use a little attention

File: actionpack/lib/action_controller/metal/instrumentation.rb (hunk 4 above)
Comment (rafaelfranca) [inline actionpack/lib/action_controller/metal/instrumentation.rb]: :+1:

File: actionpack/test/controller/filters_test.rb (hunk 5 above)
Comment (rafaelfranca) [inline actionpack/test/controller/filters_test.rb]: In this case I'd rename the test file and the test class

File: guides/source/action_controller_overview.md (hunk 7 above)
Comment (rafaelfranca) [inline guides/source/action_controller_overview.md]: I like how action callbacks sound. :+1:

File: actionpack/test/controller/filters_test.rb (hunk 5 above)
Comment (bf4) [inline actionpack/test/controller/filters_test.rb]: @rafaelfranca you mean that I should omit the comment and make a followup PR or omit the comment and change the file from filters_test.rb to callbacks_test.rb ?

File: guides/source/action_controller_overview.md (hunk 7 above)
Comment (bf4) [inline guides/source/action_controller_overview.md]: makes it easier to grep for as well :)

File: actionpack/test/controller/filters_test.rb (hunk 5 above)
Comment (rafaelfranca) [inline actionpack/test/controller/filters_test.rb]: We can do in this same PR but in a different commit.

Comment (bf4) [inline actionpack/test/controller/filters_test.rb]: Done.  Going to review for readability now.

File: actionpack/test/controller/action_callbacks_test.rb [hunk 8]
```diff
@@ -9,13 +9,13 @@ class << self
     end
 
     def before_actions
-      filters = _process_action_callbacks.select { |c| c.kind == :before }
-      filters.map!(&:raw_filter)
+      action_callbacks = _process_action_callbacks.select { |c| c.kind == :before }
+      action_callbacks.map!(&:raw_action_callback)
```
Comment (bf4) [inline actionpack/test/controller/action_callbacks_test.rb]: Replaced 'filter' with 'action_callback' throughout this file and would like to confirm it reads sensibly.

File: guides/source/action_controller_overview.md [hunk 9]
```diff
@@ -703,25 +703,25 @@ class ApplicationController < ActionController::Base
...
 class LoginsController < ApplicationController
   skip_before_action :require_login, only: [:new, :create]
 end
 ```
 
-Now, the `LoginsController`'s `new` and `create` actions will work as before without requiring the user to be logged in. The `:only` option is used to skip this filter only for these actions, and there is also an `:except` option which works the other way. These options can be used when adding filters too, so you can add a filter which only runs for selected actions in the first place.
+Now, the `LoginsController`'s `new` and `create` actions will work as before without requiring the user to be logged in. The `:only` option is used to skip this action callback only for these actions, and there is also an `:except` option which works the other way. These options can be used when adding action callbacks too, so you can add a action callback which only runs for selected actions in the first place.
 
-### After Filters and Around Filters
+### After Action Callbacks and Around Action Callbacks
```
Comment (bf4) [inline guides/source/action_controller_overview.md]: 👀

**Review by bf4 — COMMENTED** [review]
s/filter/action callback

File: guides/source/action_controller_overview.md [hunk 10]
```diff
@@ -679,14 +679,14 @@ end
 
 You may notice in the above code that we're using `render xml: @users`, not `render xml: @users.to_xml`. If the object is not a String, then Rails will automatically invoke `to_xml` for us.
 
-Filters
+Action Callbacks
```
Comment (rafaelfranca) [inline guides/source/action_controller_overview.md]: Callbacks here is lowercased since it is not an special thing.

File: guides/source/action_controller_overview.md (hunk 9 above)
Comment (rafaelfranca) [inline guides/source/action_controller_overview.md]: I'd keep both action and callbacks lowercased

**Review by rafaelfranca — COMMENTED** [review]
@fxn @matthewd WDYT?

File: actionpack/test/controller/action_callbacks_test.rb (hunk 8 above)
Comment (bf4) [inline actionpack/test/controller/action_callbacks_test.rb]: >      # NOTE: Even though we no longer refer to controller callbacks
>      # as 'filters', they are still called filters in ActiveSupport::Callbacks
>      action_callbacks.map!(&:raw_filter)

File: guides/source/5_1_release_notes.md [hunk 11]
```diff
@@ -359,7 +359,7 @@ Please refer to the [Changelog][action-mailer] for detailed changes.
 *   Allowed passing lambdas as values to the `default` method.
     ([Commit](https://github.com/rails/rails/commit/1cec84ad2ddd843484ed40b1eb7492063ce71baf))
 
-*   Added support for parameterized invocation of mailers to share before filters and defaults
+*   Added support for parameterized invocation of mailers to share before action callbacks and defaults
```
Comment (matthewd) [inline guides/source/5_1_release_notes.md]: Drop the 'before': it's falsely narrowing what we imply will work, and "before action callbacks" is a pretty unpleasant noun-phrase.

File: guides/source/action_controller_overview.md [hunk 12]
```diff
@@ -63,7 +63,7 @@ The [Layouts & Rendering Guide](layouts_and_rendering.html) explains this in mor
 
 `ApplicationController` inherits from `ActionController::Base`, which defines a number of helpful methods. This guide will cover some of these, but if you're curious to see what's in there, you can see all of them in the [API documentation](http://api.rubyonrails.org/classes/ActionController.html) or in the source itself.
 
-Only public methods are callable as actions. It is a best practice to lower the visibility of methods (with `private` or `protected`) which are not intended to be actions, like auxiliary methods or filters.
+Only public methods are callable as actions. It is a best practice to lower the visibility of methods (with `private` or `protected`) which are not intended to be actions, like auxiliary methods or action callbacks.
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: "callback handlers"? I know we later say the callback *is* the method, but strictly speaking it's not. Also again avoids putting "actions" next to "action callbacks".

File: guides/source/action_controller_overview.md [hunk 13]
```diff
@@ -679,14 +679,14 @@ end
 
 You may notice in the above code that we're using `render xml: @users`, not `render xml: @users.to_xml`. If the object is not a String, then Rails will automatically invoke `to_xml` for us.
 
-Filters
+Action callbacks
 -------
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: +`---------`

**Comment by bf4** [comment]
Test failures are unrelated https://travis-ci.org/rails/rails/builds/225056105

File: guides/source/action_controller_overview.md [hunk 14]
```diff
@@ -679,14 +679,14 @@ end
...
 
-Filters are methods that are run "before", "after" or "around" a controller action.
+Action callbacks are methods that are run "before", "after" or "around" a controller action.
 
-Filters are inherited, so if you set a filter on `ApplicationController`, it will be run on every controller in your application.
+Action callbacks are inherited, so if you set an action callback on `ApplicationController`, it will be run on every controller in your application.
 
-"before" filters may halt the request cycle. A common "before" filter is one which requires that a user is logged in for an action to be run. You can define the filter method this way:
+"before" action callbacks may halt the request cycle. A common "before" action callback is one which requires that a user is logged in for an action to be run. You can define the action callback method this way:
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: For here and each below, I think '`foo_action` callback[s]' will read better than '"foo" action callback[s]'. It was harder before, because there was no accompanying noun, and you can't really pluralize "`before_filter`s"; now that we have the "callback" there, let's use it.

For the last sentence, "callback method" seems sufficient.

File: guides/source/action_controller_overview.md [hunk 15]
```diff
@@ -703,25 +703,25 @@ class ApplicationController < ActionController::Base
 end
 ```
 
-The method simply stores an error message in the flash and redirects to the login form if the user is not logged in. If a "before" filter renders or redirects, the action will not run. If there are additional filters scheduled to run after that filter, they are also cancelled.
+The method simply stores an error message in the flash and redirects to the login form if the user is not logged in. If a "before" action callback renders or redirects, the action will not run. If there are additional action callbacks scheduled to run after that action callback, they are also cancelled.
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: "[..] additional callbacks scheduled to run after that one, [..]"

File: guides/source/action_controller_overview.md [hunk 16]
```diff
@@ -703,25 +703,25 @@ class ApplicationController < ActionController::Base
...
 
-In this example the filter is added to `ApplicationController` and thus all controllers in the application inherit it. This will make everything in the application require the user to be logged in in order to use it. For obvious reasons (the user wouldn't be able to log in in the first place!), not all controllers or actions should require this. You can prevent this filter from running before particular actions with `skip_before_action`:
+In this example the action callback is added to `ApplicationController` and thus all controllers in the application inherit it. This will make everything in the application require the user to be logged in in order to use it. For obvious reasons (the user wouldn't be able to log in in the first place!), not all controllers or actions should require this. You can prevent this action callback from running before particular actions with `skip_before_action`:
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: callback, callback

File: guides/source/action_controller_overview.md [hunk 17]
```diff
@@ -703,25 +703,25 @@ class ApplicationController < ActionController::Base
...
 
 ```ruby
 class LoginsController < ApplicationController
   skip_before_action :require_login, only: [:new, :create]
 end
 ```
 
-Now, the `LoginsController`'s `new` and `create` actions will work as before without requiring the user to be logged in. The `:only` option is used to skip this filter only for these actions, and there is also an `:except` option which works the other way. These options can be used when adding filters too, so you can add a filter which only runs for selected actions in the first place.
+Now, the `LoginsController`'s `new` and `create` actions will work as before without requiring the user to be logged in. The `:only` option is used to skip this action callback only for these actions, and there is also an `:except` option which works the other way. These options can be used when adding action callbacks too, so you can add a action callback which only runs for selected actions in the first place.
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: etc.

Basically: "action callback" on introduction and to avoid ambiguity, but otherwise "callback" on further use; it is a callback that occurs on an action, not a separate "action callback" concept fundamentally distinct from a 'regular' callback.

File: guides/source/action_controller_overview.md [hunk 18]
```diff
@@ -703,25 +703,25 @@ class ApplicationController < ActionController::Base
...
 
-### After Filters and Around Filters
+### After action callbacks and Around action callbacks
```
Comment (matthewd) [inline guides/source/action_controller_overview.md]: `###` headings seem to be titlecased elsewhere

File: actionpack/test/controller/action_callbacks_test.rb [hunk 19]
```diff
@@ -9,8 +9,10 @@ class << self
     end
 
     def before_actions
-      filters = _process_action_callbacks.select { |c| c.kind == :before }
-      filters.map!(&:raw_filter)
+      action_callbacks = _process_action_callbacks.select { |c| c.kind == :before }
+      # NOTE: Even though we no longer refer to controller callbacks
+      # as 'filters', they are still called filters in ActiveSupport::Callbacks
```
Comment (bf4) [inline actionpack/test/controller/action_callbacks_test.rb]: well, I added the note after I renamed the `raw_filter` to`raw_callback`, then had to go find where it was coming from.... `¯\_(ツ)_/¯`

**Comment by bf4** [comment]
I made the specific changes @matthewd commented on, but may have missed some places where readability could still be improved.

**Comment by bf4** [comment]
Oh well
